org.eclipse.xtext/src/org/eclipse/xtext/xtext/XtextFormatterJava.java

```diff
@@ -0,0 +1,207 @@
+				doc.prepend(element, it -> it.setNewLines(0, 0, 1));
+				if (!(element instanceof UntilToken) && !(element instanceof Group))
+					doc.surround(element, it -> it.oneSpace());
+			} else if (!first && last) {
+				doc.prepend(element, it -> it.setNewLines(0, 0, 1));
+			}
+			first = false;
+			doc.format(element);
+		}
+	}
+
+	protected void format(Wildcard wildcard, IFormattableDocument doc) {
+		doc.surround(regionFor(wildcard).keyword(":"), it -> it.oneSpace());
+	}
+
+	protected void format(CharacterRange range, IFormattableDocument doc) {
+		doc.surround(regionFor(range).keyword(".."), it -> it.noSpace());
+		formatCardinality(range, doc);
+		doc.format(range.getLeft());
+		doc.format(range.getRight());
+	}
+
+	protected void format(RuleCall call, IFormattableDocument doc) {
+		doc.append(call, it -> it.autowrap());
+		doc.prepend(regionFor(call).keyword("<"), it -> it.noSpace());
+		call.getArguments().forEach(a -> doc.format(a));
+	}
+
+	protected void format(Keyword keyword, IFormattableDocument doc) {
+		doc.surround(keyword, it -> it.autowrap());
+		formatCardinality(keyword, doc);
+	}
+
+	protected void format(NegatedToken token, IFormattableDocument doc) {
+		doc.append(regionFor(token).keyword("!"), it -> it.noSpace());
+		formatParens(token, doc);
+		formatCardinality(token, doc);
+		doc.format(token.getTerminal());
+	}
+
+	protected void format(UntilToken token, IFormattableDocument doc) {
+		doc.surround(regionFor(token).keyword("->"), it -> it.noSpace());
+		doc.format(token.getTerminal());
+	}
+
+	protected void format(Action action, IFormattableDocument doc) {
+		doc.append(regionFor(action).keyword("{"), it -> it.noSpace());
+		doc.prepend(regionFor(action).keyword("}"), it -> it.noSpace());
+	}
+
+	protected void format(CrossReference ref, IFormattableDocument doc) {
+		doc.prepend(regionFor(ref).keyword("["), it -> it.autowrap());
+		doc.append(regionFor(ref).keyword("["), it -> it.noSpace());
+		doc.prepend(regionFor(ref).keyword("]"), it -> it.noSpace());
+		doc.append(regionFor(ref).keyword("]"), it -> it.autowrap());
+		doc.surround(regionFor(ref).keyword("|"), it -> it.noSpace());
+		doc.format(ref.getType());
+	}
+
+	protected void format(Parameter param, IFormattableDocument doc) {
+		doc.surround(param, it -> it.noSpace());
+	}
+
+	protected void format(NamedArgument param, IFormattableDocument doc) {
+		doc.surround(param, it -> it.noSpace());
+		regionFor(param).keywords("=").forEach(k -> doc.surround(k, it -> it.noSpace()));
+	}
+
+	protected void format(TypeRef ref, IFormattableDocument doc) {
+		doc.surround(regionFor(ref).keyword("::"), it -> it.noSpace());
+	}
+
+	protected void format(Annotation annotation, IFormattableDocument doc) {
+		doc.surround(regionFor(annotation).keyword("@"), it -> it.noSpace());
+		doc.append(annotation, it -> it.newLine());
+	}
+
+	private void formatRule(AbstractRule rule, IFormattableDocument doc) {
+		doc.surround(regionFor(rule).keyword("returns"), it -> it.oneSpace());
```

Comment:
Is `regionFor(rule)` a cheap call? Or should we store the region in a local var?